### PR TITLE
rewrite convert

### PR DESCRIPTION
**Pull request by jackdotink**
This PR rewrites all of `convert.rs` and makes some modifications elsewhere. This allows us to properly check unbounded recursive types, get the minimum and maximum size of types for unreliable size checks, and will be easier to work with in the future.

### CODE REVIEW

File: zap/src/config.rs [hunk 1]
```diff
@@ -141,22 +192,41 @@ pub enum Enum<'src> {
...
-						size += ty_size;
+					if let Some(ty_max) = ty_max {
+						if let Some(current_max) = max {
+							if ty_max > current_max {
+								max = Some(ty_max);
+							}
+						}
 					} else {
-						return None;
+						max = None;
```
Comment (sasial-dev) [inline zap/src/config.rs:225]: Setting this to max may overwrite the previously defined Max type. Not sure if it's intended.

File: zap/src/parser/convert.rs [hunk 2]
```diff
@@ -1,142 +1,97 @@
...
 
-		for decl in self.decls {
-			match decl {
-				SyntaxDecl::Ty(tydecl) => tydecls.push(tydecl.into_config(state)),
-				SyntaxDecl::Ev(evdecl) => evdecls.push(evdecl.into_config(state)),
+		let write_checks = self.boolean_opt("write_checks", true, &config.opts).0;
+		let typescript = self.boolean_opt("typescript", false, &config.opts).0;
+
+		let server_output = self.str_opt("server_output", "network/server.lua", &config.opts).0;
+		let client_output = self.str_opt("client_output", "network/client.lua", &config.opts).0;
```
Comment (sasial-dev) [inline zap/src/parser/convert.rs:74]:
```suggestion
		let (write_checks, _) = self.boolean_opt("write_checks", true, &config.opts);
		let (typescript, _) = self.boolean_opt("typescript", false, &config.opts);

		let (server_output, _) = self.str_opt("server_output", "network/server.lua", &config.opts);
		let (client_output, _) = self.str_opt("client_output", "network/client.lua", &config.opts);
```

File: zap/src/parser/convert.rs [hunk 3]
```diff
@@ -166,228 +204,326 @@ impl<'src> SyntaxEvDecl<'src> {
...
+				}
+			}
+
+			SyntaxTyKind::Ref(ref_ty) => {
+				let ref_name = ref_ty.name;
+
+				match ref_name {
+					ref_name if ref_name == name => Some(*ref_ty),
+
+					"boolean" | "Color3" | "Vector3" | "AlignedCFrame" | "CFrame" | "unknown" => None,
```
Comment (sasial-dev) [inline zap/src/parser/convert.rs:382]: Should we put these somewhere else so we don't have to hardcode primatives in two places?

File: zap/src/parser/reports.rs [hunk 4]
```diff
@@ -227,6 +235,15 @@ impl<'src> Report<'src> {
 			Self::AnalyzeInvalidOptionalType { span, .. } => {
 				vec![Label::primary((), span.clone()).with_message("must be removed")]
 			}
+
+			Self::AnalyzeUnboundedRecursiveType {
+				decl_span, use_span, ..
+			} => {
+				vec![
+					Label::primary((), decl_span.clone()).with_message("declared here"),
+					Label::secondary((), use_span.clone()).with_message("used recursively here"),
```
Comment (sasial-dev) [inline zap/src/parser/reports.rs:244]: Where it is used recursively should be the primary issue
```suggestion
					Label::secondary((), decl_span.clone()).with_message("declared here"),
					Label::primary((), use_span.clone()).with_message("used recursively here"),
```

File: zap/src/parser/reports.rs [hunk 5]
```diff
@@ -244,7 +261,7 @@ impl<'src> Report<'src> {
 				format!("all unreliable events must be under {max_size} bytes in size"),
 				"consider adding a upper limit to any arrays or strings".to_string(),
 				"upper limits can be added for arrays by doing `[..10]`".to_string(),
-				"upper limits can be added for strings by doing `[..10]`".to_string(),
+				"upper limits can be added for strings by doing `(..10)`".to_string(),
```
Comment (sasial-dev) [inline zap/src/parser/reports.rs:264]: Maybe we should have this shows the upper limit syntax for the correct type

File: zap/src/parser/reports.rs [hunk 6]
```diff
@@ -273,7 +290,13 @@ impl<'src> Report<'src> {
 				format!("(inclusive) max: {}", max),
 			]),
 			Self::AnalyzeInvalidOptionalType { .. } => Some(vec![
-				"this type can only be used when it is not marked as optional".to_string(),
+				"you cannot have 'double optional' types, where a type is optional twice".to_string(),
+				"maps cannot have optional keys or values, as those are impossible to represent in luau".to_string(),
+				"additionally the `unknown` type cannot be optional".to_string(),
```
Comment (sasial-dev) [inline zap/src/parser/reports.rs:295]: again we should use the props to show the 'correct' one here

**Review by sasial-dev — CHANGES_REQUESTED** [review]
Looking pretty good

File: zap/src/parser/convert.rs (hunk 3 above)
Comment (jackdotink) [reply zap/src/parser/convert.rs:382]: I'd like that, but idk what to do for it

File: zap/src/parser/reports.rs (hunk 5 above)
Comment (jackdotink) [reply zap/src/parser/reports.rs:264]: potentially, but out of scope for this PR

File: zap/src/parser/reports.rs (hunk 6 above)
Comment (jackdotink) [reply zap/src/parser/reports.rs:295]: also out of scope for this PR

File: zap/src/config.rs [hunk 7]
```diff
@@ -95,37 +98,85 @@ pub enum Ty<'src> {
...
+					let tydecl = tydecls.get(name).unwrap();
+
+					tydecl.ty.size(tydecls, recursed)
 				}
 			}
+
+			Self::Enum(enum_ty) => enum_ty.size(tydecls, recursed),
+			Self::Struct(struct_ty) => struct_ty.size(tydecls, recursed),
+
+			Self::Instance(_) => (0, Some(0)),
```
Comment (sasial-dev) [inline zap/src/config.rs:172]: Should we add a comment explaining why it's `Some(0)` over `None` or do you think it's selfexplainatory?

File: zap/src/parser/convert.rs [hunk 8]
```diff
@@ -1,142 +1,97 @@
...
 
-		for decl in self.decls {
-			match decl {
-				SyntaxDecl::Ty(tydecl) => tydecls.push(tydecl.into_config(state)),
-				SyntaxDecl::Ev(evdecl) => evdecls.push(evdecl.into_config(state)),
+		let (write_checks, _) = self.boolean_opt("write_checks", true, &config.opts);
+		let (typescript, _) = self.boolean_opt("typescript", false, &config.opts);
+
+		let (server_output, _) = self.str_opt("server_output", "network/server.lua", &config.opts);
+		let (client_output, _) = self.str_opt("client_output", "network/client.lua", &config.opts);
```
Comment (sasial-dev) [inline zap/src/parser/convert.rs:74]: Let's make this even more explicit

```suggestion
		let (write_checks, ..) = self.boolean_opt("write_checks", true, &config.opts);
		let (typescript, ..) = self.boolean_opt("typescript", false, &config.opts);

		let (server_output, ..) = self.str_opt("server_output", "network/server.lua", &config.opts);
		let (client_output, ..) = self.str_opt("client_output", "network/client.lua", &config.opts);
```

File: zap/src/parser/convert.rs [hunk 9]
```diff
@@ -166,228 +204,326 @@ impl<'src> SyntaxEvDecl<'src> {
...
+					.map(|variant| {
+						if variant.1.fields.iter().any(|field| field.0.name == tag_name) {
+							self.report(Report::AnalyzeEnumTagUsed {
+								tag_span: tag.span(),
+								used_span: variant.0.span(),
+								tag: tag_name,
+							});
+						}
+
+						(variant.0.name, self.struct_ty(&variant.1))
```
Comment (sasial-dev) [inline zap/src/parser/convert.rs:336]: Let's not use `varient.x` -> move to a named tuple such as `.map(|(key, value)|`

**Review by sasial-dev — CHANGES_REQUESTED** [review]

File: zap/src/config.rs (hunk 7 above)
Comment (jackdotink) [reply zap/src/config.rs:172]: it actually shouldn't be zero, I'll fix it

File: zap/src/parser/convert.rs [hunk 10]
```diff
@@ -146,17 +101,103 @@ impl<'src> SyntaxConfig<'src> {
...
+			let event_id_size = NumTy::from_f64(1.0, self.evdecls.len() as f64).size();
+
+			// We subtract two for the `inst` array.
+			let max_unreliable_size = 900 - event_id_size - 2;
+
+			if min > max_unreliable_size {
+				self.report(Report::AnalyzeOversizeUnreliable {
+					ev_span: evdecl.span(),
+					ty_span: evdecl.data.span(),
+					max_size: 900 - event_id_size,
```
Comment (sasial-dev) [inline zap/src/parser/convert.rs:190]:
```suggestion
					max_size: max_unreliable_size,
```

File: zap/src/parser/convert.rs [hunk 11]
```diff
@@ -146,17 +101,103 @@ impl<'src> SyntaxConfig<'src> {
...
+					size: min,
+				});
+			} else if !max.is_some_and(|max| max < max_unreliable_size) {
+				self.report(Report::AnalyzePotentiallyOversizeUnreliable {
+					ev_span: evdecl.span(),
+					ty_span: evdecl.data.span(),
+					max_size: 900 - event_id_size,
```
Comment (sasial-dev) [inline zap/src/parser/convert.rs:197]:
```suggestion
					max_size: max_unreliable_size,
```

**Review by sasial-dev — CHANGES_REQUESTED** [review]
More things - then we'll be ready to merge

File: zap/src/parser/convert.rs (hunk 3 above)
Comment (sasial-dev) [reply zap/src/parser/convert.rs:385]: This is a terrible idea please don't

**Review by sasial-dev — APPROVED** [review]
It's time.

**Pull request merged**